### PR TITLE
feat: implement `Filecoin.ChainGetMessagesInTipset` API (server side)

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As part of https://github.com/ChainSafe/forest/issues/3639

Changes introduced in this pull request:

- implement Filecoin.ChainGetMessagesInTipset API (server side)

Manually tested on calibnet

request payload:
```json
{
    "jsonrpc": "2.0",
    "id": 2,
    "method": "Filecoin.ChainGetMessagesInTipset",
    "params": [
        [
            {
                "/": "bafy2bzacecctsgyvam3fxhppheh2ubkarpgjvkn7cyvesrmhttuusuwzspk6o"
            },
            {
                "/": "bafy2bzacedt2daxg7mcvbv5sdg47zruj2ogx2wxzrjidoiwpcmyzzyfazjtcu"
            },
            {
                "/": "bafy2bzaceavd2unelg7g2mghiaq2te3qkfyghdjapdyr6ltjti7dlglhuqsni"
            },
            {
                "/": "bafy2bzacebr63zjyie3oqjj7gjxr6mwifazolv44d764hc6g7mhhdehlkgel6"
            }
        ]
    ]
}
```

Lotus response
```json
{
    "jsonrpc": "2.0",
    "result": [
        {
            "Cid": {
                "/": "bafy2bzaceatwgj5tnix7n2rb3d5vig23wjypzw577rkl4tuq7kxkanbos55yk"
            },
            "Message": {
                "Version": 0,
                "To": "t01022",
                "From": "t3vyfz2c4sc7y3aaujg27cujdirvowx76fzdbowx6tsbrpps54jdgjdz5hnbgghsro7xnodtx3lszg5avmarqq",
                "Nonce": 6813,
                "Value": "147457681006493397129",
                "GasLimit": 34956948,
                "GasFeeCap": "2002463144",
                "GasPremium": "100075",
                "Method": 28,
                "Params": "gYGHCBkIftgqWCkAAYLiA4HoAiComSsWZz1fG4np8VTIbUaUSBVdNw4Li0+ANU5FrokFIRoAEDeUgBoAJ+Ek9g==",
                "CID": {
                    "/": "bafy2bzaceatwgj5tnix7n2rb3d5vig23wjypzw577rkl4tuq7kxkanbos55yk"
                }
            }
        },
        {
            "Cid": {
                "/": "bafy2bzacecvpatko6ylteuqk4prczhpmei3haxrvx6c5lkcsg4zbicqwendci"
            },
            "Message": {
                "Version": 0,
                "To": "t410fgej3qpgoyofbrx4tn4yss7pesbef26zolsew5cq",
                "From": "t410fv6ms7pamek6jigrdfrr5ygymbtkxfukfn2aa4ly",
                "Nonce": 89630,
                "Value": "0",
                "GasLimit": 8247900,
                "GasFeeCap": "1500000200",
                "GasPremium": "1500000000",
                "Method": 3844450837,
                "Params": "WIRVNdv2AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAO2JhZnliZWliem40bzcydXAzdDRuZjdjcmhhY3NnZ3RyY3RzcWJ3emtubDZvYWpnajV5Z25lanBnbWF5AAAAAAA=",
                "CID": {
                    "/": "bafy2bzacebynmaklimwgrrjlwjsvqpg57bjeqfeapvj2ohlm6e446abqrumb4"
                }
            }
        }
    ],
    "id": 2
}
```

Forest response
```json
{
    "jsonrpc": "2.0",
    "result": [
        {
            "Cid": {
                "/": "bafy2bzaceatwgj5tnix7n2rb3d5vig23wjypzw577rkl4tuq7kxkanbos55yk"
            },
            "Message": {
                "Version": 0,
                "To": "t01022",
                "From": "t3vyfz2c4sc7y3aaujg27cujdirvowx76fzdbowx6tsbrpps54jdgjdz5hnbgghsro7xnodtx3lszg5avmarqq",
                "Nonce": 6813,
                "Value": "147457681006493397129",
                "GasLimit": 34956948,
                "GasFeeCap": "2002463144",
                "GasPremium": "100075",
                "Method": 28,
                "Params": "gYGHCBkIftgqWCkAAYLiA4HoAiComSsWZz1fG4np8VTIbUaUSBVdNw4Li0+ANU5FrokFIRoAEDeUgBoAJ+Ek9g==",
                "CID": {
                    "/": "bafy2bzaceatwgj5tnix7n2rb3d5vig23wjypzw577rkl4tuq7kxkanbos55yk"
                }
            }
        },
        {
            "Cid": {
                "/": "bafy2bzacecvpatko6ylteuqk4prczhpmei3haxrvx6c5lkcsg4zbicqwendci"
            },
            "Message": {
                "Version": 0,
                "To": "t410fgej3qpgoyofbrx4tn4yss7pesbef26zolsew5cq",
                "From": "t410fv6ms7pamek6jigrdfrr5ygymbtkxfukfn2aa4ly",
                "Nonce": 89630,
                "Value": "0",
                "GasLimit": 8247900,
                "GasFeeCap": "1500000200",
                "GasPremium": "1500000000",
                "Method": 3844450837,
                "Params": "WIRVNdv2AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAO2JhZnliZWliem40bzcydXAzdDRuZjdjcmhhY3NnZ3RyY3RzcWJ3emtubDZvYWpnajV5Z25lanBnbWF5AAAAAAA=",
                "CID": {
                    "/": "bafy2bzacebynmaklimwgrrjlwjsvqpg57bjeqfeapvj2ohlm6e446abqrumb4"
                }
            }
        }
    ],
    "id": 2
}
```

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by hanabi1224** [comment]
@lemmih Do we need to write tests for each single API?

**Comment by lemmih** [comment]
> @lemmih Do we need to write tests for each single API?

Yes.

**Comment by lemmih** [comment]
> > @lemmih Do we need to write tests for each single API?
> 
> Yes.

Testing this method is easy since it has to be identical to Lotus. Add the tests in `api_cmd.rs`.

**Comment by hanabi1224** [comment]
> Testing this method is easy since it has to be identical to Lotus. Add the tests in `api_cmd.rs`.

@lemmih It seems that this API has been covered in `api_cmd.rs`, below is what I get. 

```console
     Running `target/debug/forest-tool api compare /home/me/fr/snapshots/calibnet/1063920_2023_11_06T04_13_00Z.car.zst`
| RPC Method                            | Forest              | Lotus         |
|---------------------------------------|---------------------|---------------|
| Filecoin.ChainGetBlock                | Valid               | Valid         |
| Filecoin.ChainGetGenesis              | Valid               | Valid         |
| Filecoin.ChainGetMessage (89)         | InternalServerError | Valid         |
| Filecoin.ChainGetMessage (2)          | Valid               | Valid         |
| Filecoin.ChainGetMessagesInTipset     | Valid               | Valid         |
| Filecoin.ChainGetParentMessages       | MissingMethod       | Valid         |
| Filecoin.ChainGetTipSetByHeight       | Valid               | Valid         |
| Filecoin.ChainHead                    | Valid               | Valid         |
| Filecoin.ChainReadObj                 | InvalidResponse     | Valid         |
| Filecoin.Discover                     | MissingMethod       | Valid         |
| Filecoin.MpoolPending                 | Valid               | Valid         |
| Filecoin.NetAddrsListen               | Valid               | Valid         |
| Filecoin.NetInfo                      | Valid               | MissingMethod |
| Filecoin.NetPeers                     | Valid               | Valid         |
| Filecoin.Session                      | MissingMethod       | Valid         |
| Filecoin.StartTime                    | Valid               | Valid         |
| Filecoin.StateGetActor                | Valid               | Valid         |
| Filecoin.StateGetRandomnessFromBeacon | MissingMethod       | Valid         |
| Filecoin.StateMinerActiveSectors      | MissingMethod       | Valid         |
| Filecoin.StateMinerPower (79)         | MissingMethod       | Valid         |
| Filecoin.StateNetworkName             | Valid               | Valid         |
| Filecoin.StateReadState               | MissingMethod       | Valid         |
| Filecoin.Version                      | Valid               | Valid         |
```

**Comment by lemmih** [comment]
> > Testing this method is easy since it has to be identical to Lotus. Add the tests in `api_cmd.rs`.
> 
> @lemmih It seems that this API has been covered in `api_cmd.rs`, below is what I get.
> 
> ```
>      Running `target/debug/forest-tool api compare /home/me/fr/snapshots/calibnet/1063920_2023_11_06T04_13_00Z.car.zst`
> | RPC Method                            | Forest              | Lotus         |
> |---------------------------------------|---------------------|---------------|
> | Filecoin.ChainGetBlock                | Valid               | Valid         |
> | Filecoin.ChainGetGenesis              | Valid               | Valid         |
> | Filecoin.ChainGetMessage (89)         | InternalServerError | Valid         |
> | Filecoin.ChainGetMessage (2)          | Valid               | Valid         |
> | Filecoin.ChainGetMessagesInTipset     | Valid               | Valid         |
> | Filecoin.ChainGetParentMessages       | MissingMethod       | Valid         |
> | Filecoin.ChainGetTipSetByHeight       | Valid               | Valid         |
> | Filecoin.ChainHead                    | Valid               | Valid         |
> | Filecoin.ChainReadObj                 | InvalidResponse     | Valid         |
> | Filecoin.Discover                     | MissingMethod       | Valid         |
> | Filecoin.MpoolPending                 | Valid               | Valid         |
> | Filecoin.NetAddrsListen               | Valid               | Valid         |
> | Filecoin.NetInfo                      | Valid               | MissingMethod |
> | Filecoin.NetPeers                     | Valid               | Valid         |
> | Filecoin.Session                      | MissingMethod       | Valid         |
> | Filecoin.StartTime                    | Valid               | Valid         |
> | Filecoin.StateGetActor                | Valid               | Valid         |
> | Filecoin.StateGetRandomnessFromBeacon | MissingMethod       | Valid         |
> | Filecoin.StateMinerActiveSectors      | MissingMethod       | Valid         |
> | Filecoin.StateMinerPower (79)         | MissingMethod       | Valid         |
> | Filecoin.StateNetworkName             | Valid               | Valid         |
> | Filecoin.StateReadState               | MissingMethod       | Valid         |
> | Filecoin.Version                      | Valid               | Valid         |
> ```

Good.

File: src/rpc_api/mod.rs [hunk 1]
```diff
@@ -190,6 +192,14 @@ pub mod chain_api {
     pub const CHAIN_SET_HEAD: &str = "Filecoin.ChainSetHead";
     pub const CHAIN_GET_MIN_BASE_FEE: &str = "Filecoin.ChainGetMinBaseFee";
     pub const CHAIN_GET_MESSAGES_IN_TIPSET: &str = "Filecoin.ChainGetMessagesInTipset";
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    pub struct MessageInTipset {
+        pub cid: LotusJson<Cid>,
+        pub message: LotusJson<crate::shim::message::Message>,
+    }
```
Comment (lemmih) [inline src/rpc_api/mod.rs:201]: This is `ApiMessage`.

In general, we don't want to directly use `LotusJson` types in our API. When using our client RPC library, one should never have to unpack `LotusJson`.

**Review by lemmih — CHANGES_REQUESTED** [review]
Looks good. Replace `MessageInTipset` with `ApiMessage` and it's good to go.

File: src/rpc_api/mod.rs (hunk 1 above)
Comment (hanabi1224) [reply src/rpc_api/mod.rs]: Fixed

**Review by lemmih — APPROVED** [review]

**Review by LesnyRumcajs — COMMENTED** [review]
How is this tested? Say, I change the `chain_get_messages_in_tipset` to not return any messages, ever, with `messages.clear()` at the end. Will this be caught?

**Comment by lemmih** [comment]
> How is this tested? Say, I change the `chain_get_messages_in_tipset` to not return any messages, ever, with `messages.clear()` at the end. Will this be caught?

There are tests that will catch that (by comparing our output to that of Lotus), but they're not yet part of the CI. Running the tests in CI is out of scope for this PR.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**